R/chat.R

```diff
@@ -735,6 +747,35 @@ Chat <- R6::R6Class(
 
     has_system_prompt = function() {
       length(private$.turns) > 0 && is_system_turn(private$.turns[[1]])
+    },
+
+    complete_dangling_tool_requests = function() {
+      if (length(private$.turns) == 0) {
+        return(invisible())
+      }
+
+      last_turn <- private$.turns[[length(private$.turns)]]
+      if (last_turn@role != "assistant") {
+        return(invisible())
+      }
+
+      tool_requests <- keep(last_turn@contents, is_tool_request)
+      if (length(tool_requests) == 0) {
+        return(invisible())
+      }
+
+      tool_results <- lapply(tool_requests, function(req) {
+        ContentToolResult(
+          error = "Chat ended before the tool could be invoked.",
+          request = req
+        )
+      })
+      self$add_turn(
+        tool_results_as_turn(tool_results),
+        AssistantTurn("Acknowledged", tokens = c(0, 0, 0))
+      )
```

Comment:
I'm in favor of relaxing the user-assistant paired turns constraint, but I also understand that you probably don't want to do that in this release and that it's better to stay internally consistent.

I think this approach is clean and simple _except_ for the faked assistant turn. In my initial implementation, which I'd recommend over adding an assistant turn, was to attach the tool results to the incoming user message. 

That approach introduces a dependency on #735 (in the sense that it fixes a bug with tool results and text content being sent out of order), but that PR is almost done. I'll add a commit that takes this PR in that direction so you can see how that feels (and we can revert it if you prefer this approach).